components/de_amp/browser/test/de_amp_browsertest.cc

```diff
@@ -0,0 +1,274 @@
+
+  // Go to any page
+  const GURL simple = https_server_->GetURL(kTestHost, kTestSimpleNonAmpPage);
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), simple));
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), simple);
+
+  // Now go to an AMP page
+  https_server_.reset(new net::EmbeddedTestServer(
+      net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  const GURL landing_url = https_server_->GetURL(kTestHost, kTestCanonicalPage);
+  NavigateToURLAndWaitForRedirects(original_url, landing_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, NonHttpScheme) {
+  TogglePref(true);
+  const std::string nonHttpSchemeBody =
+      R"(
+        <html amp>\
+      <head>\
+      <link rel='canonical' 
+      href='brave://settings'>\
+      </head></html>
+      )";
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, nonHttpSchemeBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, CanonicalLinkSameAsAmpPage) {
+  TogglePref(true);
+  https_server_->RegisterRequestHandler(
+      base::BindRepeating(HandleRequest, kTestHost, kTestAmpPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, PrefOff) {
+  TogglePref(false);
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  // Doesn't get De-AMPed
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, RestorePage) {
+  TogglePref(true);
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  const GURL landing_url = https_server_->GetURL(kTestHost, kTestCanonicalPage);
+  NavigateToURLAndWaitForRedirects(original_url, landing_url);
+
+  // Close & restore tab
+  browser()->tab_strip_model()->CloseSelectedTabs();
+  Profile* profile = browser()->profile();
+  chrome::OpenWindowWithRestoredTabs(profile);
+
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), landing_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, BackAndForwardNavigatesAlsoRedirect) {
```

Comment:
back/forward shouldn't redirect, the original amp page shouldn't be in the history at all. Not sure if that's what is happening here or if the test name is just confusing. `AmpURLNotStoredInHistory` maybe?